### PR TITLE
boost: update to 1.78, actually use clang on clang builds

### PR DESCRIPTION
**Pull request by oscarfv**

### CODE REVIEW

**Comment by lazka** [comment]
for gcc:

```diff
2021-12-10T00:24:32.9831178Z --- /dev/fd/63	2021-12-10 00:24:27.000000000 +0000
2021-12-10T00:24:32.9831701Z +++ /dev/fd/62	2021-12-10 00:24:27.000000000 +0000
2021-12-10T00:24:32.9832052Z @@ -9,7 +9,6 @@
2021-12-10T00:24:32.9832562Z  /mingw64/bin/libboost_contract-mt.dll
2021-12-10T00:24:32.9833172Z  /mingw64/bin/libboost_coroutine-mt.dll
2021-12-10T00:24:32.9833705Z  /mingw64/bin/libboost_date_time-mt.dll
2021-12-10T00:24:32.9834255Z -/mingw64/bin/libboost_fiber-mt.dll
2021-12-10T00:24:32.9834796Z  /mingw64/bin/libboost_filesystem-mt.dll
2021-12-10T00:24:32.9835406Z  /mingw64/bin/libboost_graph-mt.dll
2021-12-10T00:24:32.9835973Z  /mingw64/bin/libboost_iostreams-mt.dll
2021-12-10T00:24:32.9836395Z @@ -30,9 +29,6 @@
2021-12-10T00:24:32.9836842Z  /mingw64/bin/libboost_random-mt.dll
2021-12-10T00:24:32.9837349Z  /mingw64/bin/libboost_regex-mt.dll
2021-12-10T00:24:32.9837965Z  /mingw64/bin/libboost_serialization-mt.dll
2021-12-10T00:24:32.9838563Z -/mingw64/bin/libboost_stacktrace_noop-mt.dll
2021-12-10T00:24:32.9839207Z -/mingw64/bin/libboost_stacktrace_windbg_cached-mt.dll
2021-12-10T00:24:32.9839886Z -/mingw64/bin/libboost_stacktrace_windbg-mt.dll
2021-12-10T00:24:32.9840438Z  /mingw64/bin/libboost_system-mt.dll
2021-12-10T00:24:32.9840995Z  /mingw64/bin/libboost_thread-mt.dll
2021-12-10T00:24:32.9841800Z  /mingw64/bin/libboost_timer-mt.dll
```

for clang:

```diff
2021-12-10T00:21:17.3666898Z warning: database file for 'ci' does not exist (use '-Fy' to download)
2021-12-10T00:21:24.6261870Z --- /dev/fd/63	2021-12-10 00:21:17.000000000 +0000
2021-12-10T00:21:24.6262343Z +++ /dev/fd/62	2021-12-10 00:21:17.000000000 +0000
2021-12-10T00:21:24.6262903Z @@ -2,44 +2,40 @@
2021-12-10T00:21:24.6264147Z  /clang64/bin/
2021-12-10T00:21:24.6264580Z  /clang64/bin/b2.exe
2021-12-10T00:21:24.6265015Z  /clang64/bin/bjam.exe
2021-12-10T00:21:24.6265483Z -/clang64/bin/libboost_atomic-mt.dll
2021-12-10T00:21:24.6266257Z -/clang64/bin/libboost_chrono-mt.dll
2021-12-10T00:21:24.6267919Z -/clang64/bin/libboost_container-mt.dll
2021-12-10T00:21:24.6268579Z -/clang64/bin/libboost_context-mt.dll
2021-12-10T00:21:24.6269149Z -/clang64/bin/libboost_contract-mt.dll
2021-12-10T00:21:24.6269930Z -/clang64/bin/libboost_coroutine-mt.dll
2021-12-10T00:21:24.6271074Z -/clang64/bin/libboost_date_time-mt.dll
2021-12-10T00:21:24.6272917Z -/clang64/bin/libboost_fiber-mt.dll
2021-12-10T00:21:24.6273554Z -/clang64/bin/libboost_filesystem-mt.dll
2021-12-10T00:21:24.6283112Z -/clang64/bin/libboost_graph-mt.dll
2021-12-10T00:21:24.6283671Z -/clang64/bin/libboost_iostreams-mt.dll
2021-12-10T00:21:24.6284262Z -/clang64/bin/libboost_json-mt.dll
2021-12-10T00:21:24.6284788Z -/clang64/bin/libboost_locale-mt.dll
2021-12-10T00:21:24.6285477Z -/clang64/bin/libboost_log_setup-mt.dll
2021-12-10T00:21:24.6286229Z -/clang64/bin/libboost_log-mt.dll
2021-12-10T00:21:24.6286722Z -/clang64/bin/libboost_math_c99f-mt.dll
2021-12-10T00:21:24.6287221Z -/clang64/bin/libboost_math_c99l-mt.dll
2021-12-10T00:21:24.6289055Z -/clang64/bin/libboost_math_c99-mt.dll
2021-12-10T00:21:24.6289652Z -/clang64/bin/libboost_math_tr1f-mt.dll
2021-12-10T00:21:24.6290164Z -/clang64/bin/libboost_math_tr1l-mt.dll
2021-12-10T00:21:24.6290691Z -/clang64/bin/libboost_math_tr1-mt.dll
2021-12-10T00:21:24.6291224Z -/clang64/bin/libboost_nowide-mt.dll
2021-12-10T00:21:24.6291783Z -/clang64/bin/libboost_prg_exec_monitor-mt.dll
2021-12-10T00:21:24.6292399Z -/clang64/bin/libboost_program_options-mt.dll
2021-12-10T00:21:24.6293084Z -/clang64/bin/libboost_python39-mt.dll
2021-12-10T00:21:24.6293641Z -/clang64/bin/libboost_random-mt.dll
2021-12-10T00:21:24.6294180Z -/clang64/bin/libboost_regex-mt.dll
2021-12-10T00:21:24.6294737Z -/clang64/bin/libboost_serialization-mt.dll
2021-12-10T00:21:24.6295367Z -/clang64/bin/libboost_stacktrace_noop-mt.dll
2021-12-10T00:21:24.6295966Z -/clang64/bin/libboost_stacktrace_windbg_cached-mt.dll
2021-12-10T00:21:24.6296617Z -/clang64/bin/libboost_stacktrace_windbg-mt.dll
2021-12-10T00:21:24.6297202Z -/clang64/bin/libboost_system-mt.dll
2021-12-10T00:21:24.6297747Z -/clang64/bin/libboost_thread-mt.dll
2021-12-10T00:21:24.6298318Z -/clang64/bin/libboost_timer-mt.dll
2021-12-10T00:21:24.6298845Z -/clang64/bin/libboost_type_erasure-mt.dll
2021-12-10T00:21:24.6299464Z -/clang64/bin/libboost_unit_test_framework-mt.dll
2021-12-10T00:21:24.6300008Z -/clang64/bin/libboost_wave-mt.dll
2021-12-10T00:21:24.6300613Z -/clang64/bin/libboost_wserialization-mt.dll
2021-12-10T00:21:24.6301167Z +/clang64/bin/boost_atomic-mt.dll
2021-12-10T00:21:24.6301674Z +/clang64/bin/boost_chrono-mt.dll
2021-12-10T00:21:24.6302212Z +/clang64/bin/boost_container-mt.dll
2021-12-10T00:21:24.6302722Z +/clang64/bin/boost_context-mt.dll
2021-12-10T00:21:24.6303256Z +/clang64/bin/boost_contract-mt.dll
2021-12-10T00:21:24.6303770Z +/clang64/bin/boost_coroutine-mt.dll
2021-12-10T00:21:24.6304304Z +/clang64/bin/boost_date_time-mt.dll
2021-12-10T00:21:24.6304811Z +/clang64/bin/boost_filesystem-mt.dll
2021-12-10T00:21:24.6305353Z +/clang64/bin/boost_graph-mt.dll
2021-12-10T00:21:24.6305883Z +/clang64/bin/boost_iostreams-mt.dll
2021-12-10T00:21:24.6306362Z +/clang64/bin/boost_json-mt.dll
2021-12-10T00:21:24.6306857Z +/clang64/bin/boost_locale-mt.dll
2021-12-10T00:21:24.6307657Z +/clang64/bin/boost_log_setup-mt.dll
2021-12-10T00:21:24.6308301Z +/clang64/bin/boost_log-mt.dll
2021-12-10T00:21:24.6308895Z +/clang64/bin/boost_math_c99f-mt.dll
2021-12-10T00:21:24.6309483Z +/clang64/bin/boost_math_c99l-mt.dll
2021-12-10T00:21:24.6309976Z +/clang64/bin/boost_math_c99-mt.dll
2021-12-10T00:21:24.6310424Z +/clang64/bin/boost_math_tr1f-mt.dll
2021-12-10T00:21:24.6310920Z +/clang64/bin/boost_math_tr1l-mt.dll
2021-12-10T00:21:24.6311376Z +/clang64/bin/boost_math_tr1-mt.dll
2021-12-10T00:21:24.6311874Z +/clang64/bin/boost_nowide-mt.dll
2021-12-10T00:21:24.6312418Z +/clang64/bin/boost_prg_exec_monitor-mt.dll
2021-12-10T00:21:24.6312963Z +/clang64/bin/boost_program_options-mt.dll
2021-12-10T00:21:24.6313525Z +/clang64/bin/boost_python39-mt.dll
2021-12-10T00:21:24.6314003Z +/clang64/bin/boost_random-mt.dll
2021-12-10T00:21:24.6314512Z +/clang64/bin/boost_regex-mt.dll
2021-12-10T00:21:24.6315034Z +/clang64/bin/boost_serialization-mt.dll
2021-12-10T00:21:24.6315604Z +/clang64/bin/boost_system-mt.dll
2021-12-10T00:21:24.6316118Z +/clang64/bin/boost_thread-mt.dll
2021-12-10T00:21:24.6316770Z +/clang64/bin/boost_timer-mt.dll
2021-12-10T00:21:24.6317310Z +/clang64/bin/boost_type_erasure-mt.dll
2021-12-10T00:21:24.6317860Z +/clang64/bin/boost_unit_test_framework-mt.dll
2021-12-10T00:21:24.6318442Z +/clang64/bin/boost_wave-mt.dll
2021-12-10T00:21:24.6318972Z +/clang64/bin/boost_wserialization-mt.dll
2021-12-10T00:21:24.6319510Z  /clang64/include/
2021-12-10T00:21:24.6319937Z  /clang64/include/boost/
2021-12-10T00:21:24.6320393Z  /clang64/include/boost/accumulators/
```

File: mingw-w64-boost/boost-1.74-mingw-32bit-enable-tls.patch [hunk 1]
```diff
@@ -1,16 +0,0 @@
---- boost_1_74_0/boost/config/compiler/gcc.hpp.orig	2020-10-07 19:30:01.735299600 +0300
```
Comment (lazka) [inline mingw-w64-boost/boost-1.74-mingw-32bit-enable-tls.patch:1]: why was this removed?

**Comment by lazka** [comment]
note that gcc in the clang env is just clang, so it was always using clang.

File: mingw-w64-boost/boost-1.74-mingw-32bit-enable-tls.patch (hunk 1 above)
Comment (oscarfv) [reply mingw-w64-boost/boost-1.74-mingw-32bit-enable-tls.patch:1]: It is in boost now.

**Comment by oscarfv** [comment]
> note that gcc in the clang env is just a copy of clang, so it was always using clang.

This way boost build can apply specific settings for clang, if necessary.

**Comment by oscarfv** [comment]
The changes on clang .dll names probably come from using --toolset=clang. The missing boost_fiber etc look like an upstream bug, maybe it doesn't handle the fact that g++/clang++ default to C++17 now. I'll investigate.

Thanks for the review.

**Comment by oscarfv** [comment]
This is weird. The build simply ignores `boost_fiber` and `boost_backtrace` without giving any hint. It just ends saying that the build is complete.

Created issue upstream: https://github.com/boostorg/boost/issues/593

File: mingw-w64-boost/boost-1.74-mingw-32bit-enable-tls.patch (hunk 1 above)
Comment (lazka) [reply mingw-w64-boost/boost-1.74-mingw-32bit-enable-tls.patch:1]: Ah, missed that they guarded it with ` (__GNUC__ < 11) ` now. thanks

**Comment by oscarfv** [comment]
Removing `--no-cmake-config` from the command line fixes the build up to some point: `boost_fiber` and `boost_stacktrace` are built but not copied to `stage/lib` nor installed. This can be worked around by installing them with `cp` from `install()`.

That parameter was introduced in 4e4b2d9310b74c1e58ef9addf500eea880b5fd02 with no hint about its justification, as usual. If necessary I'll simply remove the parameter and if that causes a problem to anybody, the correct action is to file a bug report and/or submit a PR with an explanation of the problem on the commit message or as a comment on the code.

**Comment by lazka** [comment]
> with no hint about its justification, as usua

there is a bit in the PR https://github.com/msys2/MINGW-packages/pull/5326, but not very useful.. @Adsun701 do you remember?

**Comment by Adsun701** [comment]
@lazka I remember that the `--no-cmake-config` option was introduced because without it, some packages that were depending on boost were failing to build with an error `"cannot detect boost libraries using cmake config"`. If the option were removed it would be possible that these packages would once again fail to build, but I'm not 100% sure.

**Comment by oscarfv** [comment]
After applying a patch from upstream to `boost.build` it is no longer necessary to remove `--no-cmake-config` for building and installing `boost_fiber` and `boost_stacktrace`.

Hope everything is ok now.

**Comment by oscarfv** [comment]
C.I. fails while checking sha256 of the patch I added, but on my machine `updpkgsums` says everything is fine :-/

**Comment by lazka** [comment]
that's usually when you commit a text file with windows line endings, and then git will normalize on checkout.

**Comment by oscarfv** [comment]
Thaks. Yes, just realized that a minute ago. Force-pushed with new sha256.

**Comment by oscarfv** [comment]
Are there any outstanding problems with this PR?

**Comment by lazka** [comment]
lgtm, thanks

**Comment by podsvirov** [comment]
> note that gcc in the clang env is just a copy of clang, so it was always using clang.

I have trubles to rebuild `boost` for `CLANG` subsystems in #10719.

```console
2022-02-09T16:47:33.0440184Z Building B2 engine..
2022-02-09T16:47:33.2473340Z 
2022-02-09T16:47:33.2478485Z A C++11 capable compiler is required for building the B2 engine.
2022-02-09T16:47:33.2479945Z Toolset 'gcc' does not appear to support C++11.
2022-02-09T16:47:33.2489330Z 
2022-02-09T16:47:33.2567851Z > g++ -x c++ -std=c++11  check_cxx11.cpp
2022-02-09T16:47:33.2790293Z ./tools/build/src/engine/build.sh: line 127: g++: command not found
2022-02-09T16:47:33.3001943Z > g++ -x c++ -std=c++11 -D_GNU_SOURCE  check_cxx11.cpp
2022-02-09T16:47:33.3227189Z ./tools/build/src/engine/build.sh: line 127: g++: command not found
2022-02-09T16:47:33.3428398Z > g++ -x c++ -std=c++11 -pthread  check_cxx11.cpp
2022-02-09T16:47:33.3659662Z ./tools/build/src/engine/build.sh: line 127: g++: command not found
2022-02-09T16:47:33.3899499Z 
2022-02-09T16:47:33.3902097Z 
2022-02-09T16:47:33.3903741Z ** Note, the C++11 capable compiler is _only_ required for building the B2
2022-02-09T16:47:33.3911331Z ** engine. The B2 build system allows for using any C++ level and any other
2022-02-09T16:47:33.3911785Z 
2022-02-09T16:47:33.3934049Z ** supported language and resource in your projects.
2022-02-09T16:47:33.3934453Z Failed to build B2 build engine
```

Looks like `gcc` copy now missing.

**Comment by filnet** [comment]
I don't know if it helps but the rust CLANG build fails with a similar error:
```
      Compiling winapi-x86_64-pc-windows-gnu v0.4.0
       Running `C:/_/mingw-w64-rust/src/CLANG64/build/x86_64-pc-windows-gnu/stage0/bin/rustc.exe --crate-name build_script_build 'C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-x86_64-pc-windows-gnu-0.4.0\build.rs' --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C embed-bitcode=no -C debuginfo=0 -C metadata=f1d3b506361404ce -C extra-filename=-f1d3b506361404ce --out-dir 'C:/_/mingw-w64-rust/src/CLANG64/build/bootstrap\debug\build\winapi-x86_64-pc-windows-gnu-f1d3b506361404ce' -L 'dependency=C:/_/mingw-w64-rust/src/CLANG64/build/bootstrap\debug\deps' --cap-lints allow -Wrust_2018_idioms -Wunused_lifetimes -Wsemicolon_in_expressions_from_macros -Dwarnings`
  error: linker `x86_64-w64-mingw32-gcc` not found
    |
    = note: The system cannot find the file specified. (os error 2)
  
  error: could not compile `winapi-x86_64-pc-windows-gnu` due to previous error
  
  Caused by:
    process didn't exit successfully: `C:/_/mingw-w64-rust/src/CLANG64/build/x86_64-pc-windows-gnu/stage0/bin/rustc.exe --crate-name build_script_build 'C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-x86_64-pc-windows-gnu-0.4.0\build.rs' --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C embed-bitcode=no -C debuginfo=0 -C metadata=f1d3b506361404ce -C extra-filename=-f1d3b506361404ce --out-dir 'C:/_/mingw-w64-rust/src/CLANG64/build/bootstrap\debug\build\winapi-x86_64-pc-windows-gnu-f1d3b506361404ce' -L 'dependency=C:/_/mingw-w64-rust/src/CLANG64/build/bootstrap\debug\deps' --cap-lints allow -Wrust_2018_idioms -Wunused_lifetimes -Wsemicolon_in_expressions_from_macros -Dwarnings` (exit code: 1)
```

Also noticed this:
```
  resolving dependencies...
  :: There are 2 providers available for mingw-w64-clang-x86_64-gcc-libs:
  :: Repository staging
     1) mingw-w64-clang-x86_64-libc++
  :: Repository clang64
     2) mingw-w64-clang-x86_64-libc++
```
